sdk/core/core/src/az_json_data.c

```diff
@@ -0,0 +1,234 @@
+  {
+    uint8_t * const p = builder->buffer.begin + builder->length;
+    AZ_RETURN_IF_FAILED(
+        az_span_builder_append_zeros(builder, _az_align_ceil(p, data.type.align) - p));
+  }
+  *out = builder->buffer.begin + builder->length;
+  return az_span_builder_append(builder, _az_data_get_span(data));
+}
+
+AZ_NODISCARD az_result
+_az_span_builder_top_aligned_append(az_span_builder * const builder, _az_data const data) {
+  size_t const size = builder->buffer.size;
+  if (size < data.type.size) {
+    return AZ_ERROR_BUFFER_OVERFLOW;
+  }
+  uint8_t * const begin = builder->buffer.begin;
+  uint8_t * const end = _az_align_floor(begin + size - data.type.size, data.type.align);
+  if (end < begin + builder->length) {
+    return AZ_ERROR_BUFFER_OVERFLOW;
+  }
+  size_t const new_size = end - begin;
+  {
+    az_mut_span result = { 0 };
+    AZ_RETURN_IF_FAILED(az_mut_span_move(
+        az_mut_span_drop(builder->buffer, new_size), _az_data_get_span(data), &result));
+  }
+  builder->buffer.size = new_size;
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result _az_span_builder_top_array_revert(
+    az_span_builder * const builder,
+    _az_type const item_type,
+    size_t const new_size,
+    void const ** const out_array_begin,
+    size_t * const out_array_size) {
+  AZ_CONTRACT_ARG_NOT_NULL(builder);
+  AZ_CONTRACT_ARG_NOT_NULL(out_array_begin);
+  AZ_CONTRACT_ARG_NOT_NULL(out_array_size);
+
+  // 1. restore buffer size
+  size_t const offset = builder->buffer.size;
+  builder->buffer.size = new_size;
+  az_mut_span const buffer = builder->buffer;
+
+  // 2. revert the array.
+  size_t const item_count = (new_size - offset) / item_type.size;
+  for (size_t i = 0; i < item_count / 2; ++i) {
+    az_mut_span const a
+        = az_mut_span_take(az_mut_span_drop(buffer, offset + i * item_type.size), item_type.size);
+    az_mut_span const b = az_mut_span_take(
+        az_mut_span_drop(buffer, offset + (item_count - i - 1) * item_type.size), item_type.size);
+    az_mut_span_swap(a, b);
+  }
+
+  // 3. move it to front.
+  az_span const result_array = az_span_take(
+      az_span_drop(az_mut_span_to_span(buffer), offset), item_count * item_type.size);
+  _az_data const data = {
+    .p = result_array.begin,
+    .type = { .size = result_array.size, .align = item_type.align },
+  };
+  AZ_RETURN_IF_FAILED(_az_span_builder_aligned_append(builder, data, out_array_begin));
+  *out_array_size = item_count;
+
+  return AZ_OK;
+}
+
+AZ_INLINE AZ_NODISCARD az_result _az_span_builder_append_json_data(
+    az_span_builder * const builder,
+    az_json_data const data,
+    az_json_data const ** const out) {
+  void const * p = { 0 };
+  AZ_RETURN_IF_FAILED(_az_span_builder_aligned_append(builder, _AZ_DATA(data), &p));
+  *out = (az_json_data const *)p;
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result _az_span_builder_write_json_string(
+    az_span_builder * const builder,
+    az_span const json_string,
+    az_span * const out) {
+  // TODO: escape symbols.
```

Comment:
Are you going to implement it?

Reply:
as a separate PR, otherwise, I will add an issue.